### PR TITLE
Part 2: JEP 360 Sealed Classes

### PR DESCRIPTION
**Pull request by theresa-m**
related to: #9626

Class loading restrictions and tests for JEP 360. JVM should throw an IncompatibleClassChangeError if a class that inherits from a sealed superclass/interface is not listed in its PermittedSubclasses attribute.

Signed-off-by: Theresa Mammarella Theresa.T.Mammarella@ibm.com

### CODE REVIEW

**Comment by theresa-m** [comment]
This will need to be merged after the latest asm upgrade: https://github.com/AdoptOpenJDK/TKG/pull/94

**Comment by pshipton** [comment]
I expect TKG is frozen until the Adopt July releases are completed.

File: runtime/vm/createramclass.cpp [hunk 1]
```diff
@@ -1577,6 +1577,42 @@ popLoadingOrLinkingStack(J9VMThread *vmThread, J9StackElement **stack, J9Pool *s
...
+	BOOLEAN result = FALSE;
+	U_32 *permittedSubclassesCountPtr = getNumberOfPermittedSubclassesPtr(superRomClass);
+
+	/* find matching subclass name */
+	for (U_32 index = 0; index < *permittedSubclassesCountPtr; index++) {
+		J9UTF8* permittedSubclassNameUtf8 = permittedSubclassesNameAtIndex(permittedSubclassesCountPtr, index);
+		U_8 *permittedSubclassName = J9UTF8_DATA(permittedSubclassNameUtf8);
+		U_16 permittedSubclassLength = J9UTF8_LENGTH(permittedSubclassNameUtf8);
+
+		if ((permittedSubclassLength == classNameLength)
```
Comment (gacholio) [inline runtime/vm/createramclass.cpp:1606]: Please use this macro instead (UTFs are not null terminated):

https://github.com/eclipse/openj9/blob/ee1f3d1ae159643f5cf8d3b3554224ee6282f082/runtime/oti/j9.h#L171

File: runtime/vm/createramclass.cpp [hunk 2]
```diff
@@ -1577,6 +1577,42 @@ popLoadingOrLinkingStack(J9VMThread *vmThread, J9StackElement **stack, J9Pool *s
...
+ * @param className name of subclass
+ * @param classNameLength length of subclass name
+ * @return TRUE if subclass can legally inherit the super, FALSE otherwise.
+ */
+static VMINLINE BOOLEAN
+isClassPermittedBySealedSuper(J9ROMClass *superRomClass, U_8* className, U_16 classNameLength)
+{
+	if (! J9ROMCLASS_IS_SEALED(superRomClass)) {
+		/* for non-sealed classes all subclasses are permitted (the final case is handled elsewhere). */
+		return TRUE;
```
Comment (gacholio) [inline runtime/vm/createramclass.cpp:1594]: Please avoid the use of multiple returns.

Comment (theresa-m) [reply runtime/vm/createramclass.cpp]: updated.

**Comment by gacholio** [comment]
jenkins test sanity,extended win jdk15

File: runtime/vm/createramclass.cpp [hunk 3]
```diff
@@ -1577,6 +1577,41 @@ popLoadingOrLinkingStack(J9VMThread *vmThread, J9StackElement **stack, J9Pool *s
...
+ * @param className name of subclass
+ * @param classNameLength length of subclass name
+ * @return TRUE if subclass can legally inherit the super, FALSE otherwise.
+ */
+static VMINLINE BOOLEAN
+isClassPermittedBySealedSuper(J9ROMClass *superRomClass, U_8* className, U_16 classNameLength)
+{
+	BOOLEAN result = FALSE;
+
+	if (! J9ROMCLASS_IS_SEALED(superRomClass)) {
```
Comment (gacholio) [inline runtime/vm/createramclass.cpp:1594]: Once testing is complete, please remove the extra blank here and I'll merge.

Comment (theresa-m) [reply runtime/vm/createramclass.cpp:1593]: I have removed the space. The tests will not pass until AdoptOpenJDK/TKG#94 is merged.

Comment (gacholio) [reply runtime/vm/createramclass.cpp:1593]: OK let me know when it's ready.

**Comment by theresa-m** [comment]
@gacholio the dependent TKG pr has been merged, this is ready to be tested/merged.

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk15

**Pull request merged**